### PR TITLE
Drawing Primitives with Gizmos

### PR DESCRIPTION
**Pull request by RobWalt**
The PR is in a reviewable state now in the sense that the basic implementations are there. There are still some ToDos that I'm aware of:

- [x] docs for all the new structs and traits
- [x] implement `Default` and derive other useful traits for the new structs
- [x] Take a look at the notes again (Do this after a first round of reviews)
- [x] Take care of the repetition in the circle drawing functions

---

# Objective

- TLDR: This PR enables us to quickly draw all the newly added primitives from `bevy_math` in immediate mode with gizmos
- Addresses #10571

## Solution

- This implements the first design idea I had that covered everything that was mentioned in the Issue https://github.com/bevyengine/bevy/issues/10571#issuecomment-1863646197

--- 

## Caveats

- I added the `Primitive(2/3)d` impls for `Direction(2/3)d` to make them work with the current solution. We could impose less strict requirements for the gizmoable objects and remove the impls afterwards if the community doesn't like the current approach.

---

## Changelog

- implement capabilities to draw ellipses on the gizmo in general (this was required to have some code which is able to draw the ellipse primitive)
- refactored circle drawing code to use the more general ellipse drawing code to keep code duplication low
- implement `Primitive2d` for `Direction2d` and impl `Primitive3d` for `Direction3d`
- implement trait to draw primitives with specialized details with gizmos
  - `GizmoPrimitive2d` for all the 2D primitives
  - `GizmoPrimitive3d` for all the 3D primitives
  - (question while writing this: Does it actually matter if we split this in 2D and 3D? I guess it could be useful in the future if we do something based on the main rendering mode even though atm it's kinda useless) 

---

### CODE REVIEW

**Comment by RobWalt** [comment]
This PR currently includes some other general changes to Gizmos which could be factored out. They are:

- Generalizing circle drawing to ellipse drawing for gizmos
- Soon ™️  Arc3d drawing support for gizmos

The prior was pretty straight-forward. For the latter, here are some references how other game engines implement Arc3d, which I used for orientation:

- unity - https://docs.unity3d.com/ScriptReference/Handles.DrawWireArc.html
- unreal blueprints - https://docs.unrealengine.com/5.2/en-US/BlueprintAPI/GeometryScript/PolyPath/CreateArcPath3D/
- godot (doesn't seem to have built in arcs, here's a lib that implements it) - https://github.com/nyxkn/godot-draw3d/blob/main/addons/draw3d/draw_3d.gd#L212-L240

**Comment by RobWalt** [comment]
Some general notes for reviewers:

- I'll try to add an example to prove everything is working fine
- If you want I can split out the new general functionalities of `Gizmos` into a separate PR (Ellipses & Arcs)

@Jondolf Can you take a look over this and review it if you find some time?

**Comment by Jondolf** [comment]
I think the 3D arc should be in a separate PR since it's new functionality that isn't related to the primitives, which are the core focus of this PR. But the ellipse might be fine since it's needed for the primitives too.

I responded about the API in the issue and added a proposal for an alternative: https://github.com/bevyengine/bevy/issues/10571#issuecomment-1883854471. I felt like it'd be too long and slightly off-topic to post here.

**Review by rodolphito — CHANGES_REQUESTED** [review]
please remove flake.lock, flake.nix, and template.nix

**Comment by RobWalt** [comment]
@rodolphito I will when this branch is ready. Atm I'm facing the problem described in https://github.com/bevyengine/bevy/issues/10571#issuecomment-1893429450 which blocks this here from a merge anyways.

**Comment by RobWalt** [comment]
Resolved the last issue with this and also removed the nix files

**Comment by nothendev** [comment]
I'm wondering if Self::Output<**'_**> is 2018 idioms..

**Comment by RobWalt** [comment]
Just so I don't forget: This needs a rebase on this https://github.com/bevyengine/bevy/pull/11540 and some adjustments because of it.

**Review by Jondolf — COMMENTED** [review]
Looks pretty good overall, but I'm not a huge fan of having the position, rotation, and especially color configured via builders. In my opinion, the API should be:

```rust
gizmos.primitive_2d(primitive, translation, rotation, color).builder_stuff();
```

I can't think of many cases where I'd want to always draw something at the origin with gizmos, and primitives by themselves don't have any position or rotation associated with them. The color should also be given explicitly, as assuming a default color for gizmos is not a good idea in my opinion.

Using a builder pattern for the positioning and color is also inconsistent: all existing gizmos have `color` as the last argument(s), and they all have some kind of position as an argument.

The current names for the builders are also inconsistent: there's `center`, `start_position`, `normal_position`, `translation`, `position`... The user basically needs to guess what the name is for each primitive. I would unify them all to be just `translation` (as an argument to the gizmo method), as it simply translates the primitive relative to the origin.

One argument against this is that we might want to define e.g. line segments with one endpoint at the start position instead of having the center of the line segment be at the origin, but... do we? If you need that representation, you could just use `gizmos.line_2d(...)`. All primitives are currently assumed to be centered at the origin, and I'm not sure if we want to break that assumption for gizmos.

The `primitives.rs` gizmo file is also getting really big now; I would split it into `primitives2d` and `primitives3d` modules, or a `primitives` module with `dim2` and `dim3` modules.

File: crates/bevy_gizmos/src/primitives.rs [hunk 1]
```diff
@@ -0,0 +1,1954 @@
...
+/// Builder for configuring the drawing options of [`Ellipse`].
+pub struct Ellipse2dBuilder<'a, 'w, 's, T: GizmoConfigGroup> {
+    gizmos: &'a mut Gizmos<'w, 's, T>,
+
+    half_width: f32,  // Half-width of the ellipse
+    half_height: f32, // Half-height of the ellipse
+
+    center: Vec2, // Position of the center of the ellipse
+    color: Color, // Color of the ellipse
+}
```
Comment (Jondolf) [inline crates/bevy_gizmos/src/primitives.rs:136]: Ellipses should support rotation like other primitives. Imo every primitive (even the ones with a direction or normal) should be rotatable, as they don't have an inherent rotation associated with them

**Comment by Jondolf** [comment]
Another reason why having `translation` and `rotation` as arguments could be useful: the primitive gizmo method is in a trait, so you could make a generic system or method that renders the given primitive at some position. If the position and rotation are in the primitive-specific builders, a generic system like this is not possible (the gizmo is always at the origin).

Example use case: bevy_xpbd has a physics debug rendering pipeline. I could convert each collider to a primitive if possible, and call the primitive gizmo method for collider visualization. If I can't do this with generics, each primitive type must be handled individually.

**Comment by Jondolf** [comment]
TLDR of what I'd personally change:
- Remove position, rotation, and color from all builders
- Change API to `gizmos.primitive_2d(primitive, translation, rotation, color)`
  - Every primitive is centered at the origin by default, and supports rotation
- Split `primitives` into 2D and 3D files
If you want to keep support for e.g. segments starting from a point instead of being centered at the origin, you could add that as a configuration option. But imo all primitives should be centered at the origin *by default*.

Sorry if this is a lot 😅 and feel free to disagree! This is just what I would find to be the most useful and consistent API

**Comment by RobWalt** [comment]
Makes sense! I implemented everything as requested. Should be more in line with the rest of the gizmo code. Another nice side effect of these changes is that the code is much "simpler". Less code = easier to maintain, so yay 🎉 

The only thing that I didn't use from the review is the `translation` name. I used `position` instead to make it even more similar to the rest of the gizmo methods.

Thanks for the review 😁 👍🏼

**Comment by Jondolf** [comment]
I suggested `translation` because I don't consider e.g. lines and planes to have a specific "position"; they can simply be translated relative to the origin. This is probably nitpicking though, so `position` is fine too :)

I made a slightly stripped down version of the `3d_gizmos` example:

https://github.com/bevyengine/bevy/assets/57632562/6942b2ad-98c6-4769-bb52-23d0948ef274

The primitives look good, but I see two things that I consider to be issues.

Firstly, I think the primitives (capsule, cylinder, cone, and conical frustum) should be upright instead of facing Z. I asked about this on the Discord, and heard that [gizmos should typically face negative Z](https://discord.com/channels/691052431525675048/730525730601041940/1201138281186852864), but I think we agreed that it's reasonable for these primitives to be upright. The primitive meshes and colliders would most certainly be upright, so making gizmos differ from this is probably not a good idea.

It's also just intuitive in my opinion; if I draw a cone with no rotation, I'd expect it to be upright like a traffic cone, not pointing towards Z or negative Z. The names of the properties of the primitives also suggest this: e.g. cylinders and cones currently have a `height` property, not length.

Secondly, in my opinion, cones and conical frusta should not start at the base, but instead be centered on the origin like other primitives (i.e. the center of the cone is at the geometric center, half-way between the tip and base). Cylinders, conical frustums, and cones can be treated as different versions of the same shape: many engines only have a cylinder, but it's actually a conical frustum since it has two radii; cylinders and cones are a special case of this. They should all be positioned similarly, centered at the origin, like other primitives.

For reference, Godot has conical frusta (called cylinder) centered like this for both the mesh and collider, Parry has cone and cylinder colliders centered, and Blender also centers cylinders (but doesn't seem to have two radii for them).

Side note: The gizmo examples cycle through primitives with the Up/Down arrow keys, but the text overlays don't say this.

**Comment by RobWalt** [comment]
Good points 👍🏼  I'll tackle these asap.

The upright thing makes total sense and I already thought about if I should make that change. I used Z up mainly for my mental model while developing since I'm used to it from other applications using mainly 2D structures with an optional Z coordinate in specific cases.

As for the centering: That's also reasonable to further unify implicit assumptions, so cool.

I'll also add UI text telling the example users how to navigate the 3D gizmos.

File: crates/bevy_gizmos/src/primitives/dim2.rs [hunk 2]
```diff
@@ -0,0 +1,434 @@
...
+
+        // plane line
+        let direction = Direction2d::new_unchecked(normal.perp());
+        [1.0, -1.0].into_iter().for_each(|sign| {
+            self.line_2d(
+                position,
+                position + sign * direction.clamp_length(1000.0, 1000.0),
+                color,
+            );
+        });
```
Comment (Jondolf) [inline crates/bevy_gizmos/src/primitives/dim2.rs:162]: Same as `Line2d`
```suggestion
        self.line_2d(
            position - direction * 1_000_000.0,
            position + direction * 1_000_000.0,
            color,
        );
```

File: crates/bevy_gizmos/src/primitives/dim2.rs [hunk 3]
```diff
@@ -0,0 +1,434 @@
...
+        rotation: Mat2,
+        color: Color,
+    ) -> Self::Output<'_> {
+        if !self.enabled {
+            return;
+        }
+
+        let start = position;
+        let end = position + (rotation * *primitive);
+        self.arrow_2d(start, end, color);
```
Comment (Jondolf) [inline crates/bevy_gizmos/src/primitives/dim2.rs:49]: This will currently draw an arrow with a length of 1 pixel with a default 2D camera. Not entirely sure what the default value *should* be, maybe 50 pixels? It looks decent, but it's quite arbitrary... Either way, the arrow length should probably be configurable.

File: crates/bevy_gizmos/src/primitives/dim3.rs [hunk 4]
```diff
@@ -0,0 +1,822 @@
...
+
+        let dir = rotation * *primitive.direction;
+        self.arrow(position, position + dir, color);
+        [1.0, -1.0].into_iter().for_each(|sign| {
+            self.line(
+                position,
+                position + sign * dir.clamp_length(1000.0, 1000.0),
+                color,
+            );
+        });
```
Comment (Jondolf) [inline crates/bevy_gizmos/src/primitives/dim3.rs:188]:
```suggestion
        self.line(position - dir * 1000.0, position + dir * 1000.0, color);
```

File: crates/bevy_gizmos/src/primitives/dim2.rs [hunk 5]
```diff
@@ -0,0 +1,434 @@
...
+    ) -> Self::Output<'_> {
+        if !self.enabled {
+            return;
+        }
+
+        let direction = rotation * *primitive.direction;
+
+        let start = position;
+        let end = position + direction;
+        self.arrow_2d(start, end, color);
```
Comment (Jondolf) [inline crates/bevy_gizmos/src/primitives/dim2.rs:119]: No strong feelings here, but it might make more sense to have the arrowhead at the center of the line, at `position`
```suggestion
        let start = position - direction;
        let end = position;
        self.arrow_2d(start, end, color);
```

File: crates/bevy_gizmos/src/primitives/dim3.rs [hunk 6]
```diff
@@ -0,0 +1,822 @@
...
+        position: Vec3,
+        rotation: Quat,
+        color: Color,
+    ) -> Self::Output<'_> {
+        if !self.enabled {
+            return;
+        }
+
+        let dir = rotation * *primitive.direction;
+        self.arrow(position, position + dir, color);
```
Comment (Jondolf) [inline crates/bevy_gizmos/src/primitives/dim3.rs:181]: Same as 2D
```suggestion
        let dir = rotation * *primitive.direction;
        self.arrow(position - dir, position, color);
```
Also, the naming and variables are slightly different from 2D (2D has `direction` and variables `start` and `end`)

File: crates/bevy_gizmos/src/primitives/dim3.rs [hunk 7]
```diff
@@ -0,0 +1,822 @@
...
+        rotation: Quat,
+        color: Color,
+    ) -> Self::Output<'_> {
+        if !self.enabled {
+            return;
+        }
+
+        let [half_extend_x, half_extend_y, half_extend_z] = primitive.half_size.to_array();
+
+        let vertices @ [a, b, c, d, e, f, g, h] = [
```
Comment (Jondolf) [inline crates/bevy_gizmos/src/primitives/dim3.rs:287]: Side note: I like this syntax, haven't seen it before 😄

File: crates/bevy_gizmos/src/primitives/dim3.rs [hunk 8]
```diff
@@ -0,0 +1,822 @@
...
+        let normal = *rotation * Vec3::Y;
+
+        [1.0, -1.0].into_iter().for_each(|sign| {
+            // use "-" here since rotation is ccw and otherwise the caps would face the wrong way
+            // around
+            let center = *position + sign * *half_length * normal;
+            let top = center + sign * *radius * normal;
+            draw_cap(gizmos, *radius, *segments, *rotation, center, top, *color);
+            draw_circle(gizmos, *radius, *segments, *rotation, center, *color);
+        });
```
Comment (Jondolf) [inline crates/bevy_gizmos/src/primitives/dim3.rs:493]: Does the comment refer to using a negative `sign` for the bottom cap? The actual code doesn't have "-" in the calculations. If so, maybe just say
```suggestion
        [1.0, -1.0].into_iter().for_each(|sign| {
            // use a negative `sign` for the bottom cap since rotation is ccw and otherwise the cap would face
            // the wrong way around
            let center = *position + sign * *half_length * normal;
            let top = center + sign * *radius * normal;
            draw_cap(gizmos, *radius, *segments, *rotation, center, top, *color);
            draw_circle(gizmos, *radius, *segments, *rotation, center, *color);
        });
```
The comment says "caps" though so it's probably referring to something else?  Or maybe it's just outdated

File: crates/bevy_gizmos/src/primitives/dim3.rs [hunk 9]
```diff
@@ -0,0 +1,822 @@
...
+pub struct Cone3dBuilder<'a, 'w, 's, T: GizmoConfigGroup> {
+    gizmos: &'a mut Gizmos<'w, 's, T>,
+
+    // Radius of the cone
+    radius: f32,
+    // Height of the cone
+    height: f32,
+
+    // Center of the base of the cone
+    position: Vec3,
```
Comment (Jondolf) [inline crates/bevy_gizmos/src/primitives/dim3.rs:519]: Outdated
```suggestion
    // Center of the cone, half-way between the tip and base
    position: Vec3,
```

File: crates/bevy_gizmos/src/primitives/dim3.rs [hunk 10]
```diff
@@ -0,0 +1,822 @@
...
+
+    // Radius of the top circle
+    radius_top: f32,
+    // Radius of the bottom circle
+    radius_bottom: f32,
+    // Height of the conical frustum
+    height: f32,
+
+    // Center of the base circle of the conical frustum
+    position: Vec3,
```
Comment (Jondolf) [inline crates/bevy_gizmos/src/primitives/dim3.rs:613]: Outdated; "base" also isn't as clear for conical frusta as it is for cones
```suggestion
    // Center of the conical frustum, half-way between the top and bottom
    position: Vec3,
```

File: crates/bevy_gizmos/src/primitives/helpers.rs [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+
+// helpers - affine transform
+
+pub fn rotate_then_translate_2d(rotation: Mat2, position: Vec2) -> impl Fn(Vec2) -> Vec2 {
+    move |v| rotation * v + position
+}
+
+pub fn rotate_then_translate_3d(rotation: Quat, translation: Vec3) -> impl Fn(Vec3) -> Vec3 {
+    move |v| rotation * v + translation
+}
```
Comment (Jondolf) [inline crates/bevy_gizmos/src/primitives/helpers.rs:16]: These feel a bit unnecessary to me, but keeping them is probably fine

**Review by Jondolf — CHANGES_REQUESTED** [review]
Looking pretty good now! I have some final nits and fixes, mainly around lines and some comments.

You should probably also add the primitive gizmo traits to the `bevy_gizmos` prelude so that `primitive_2d` and `primitive_3d` work without having to explicitly import traits:
```rust
/// The `bevy_gizmos` prelude.
pub mod prelude {
    #[doc(hidden)]
    pub use crate::{
        aabb::{AabbGizmoConfigGroup, ShowAabbGizmo},
        config::{DefaultGizmoConfigGroup, GizmoConfig, GizmoConfigGroup, GizmoConfigStore},
        gizmos::Gizmos,
        // This
        primitives::{dim2::GizmoPrimitive2d, dim3::GizmoPrimitive3d},
        AppGizmoBuilder,
    };
}
```

**Comment by Jondolf** [comment]
FYI, `Capsule` was renamed to `Capsule3d` and a `Capsule2d` primitive was added: #11585

It should probably be added as a 2D gizmo too, it'd just be two semicircles and lines to connect them

**Comment by RobWalt** [comment]
> FYI, `Capsule` was renamed to `Capsule3d` and a `Capsule2d` primitive was added: #11585
> 
> It should probably be added as a 2D gizmo too, it'd just be two semicircles and lines to connect them

Yeah I saw that. Solid addition btw! Really cool seeing the primitives evolve. They grow up so fast! 🥲

File: crates/bevy_gizmos/src/primitives/dim2.rs (hunk 5 above)
Comment (RobWalt) [reply crates/bevy_gizmos/src/primitives/dim2.rs:119]: I think I agree almost everywhere. An exception is here. It looks a bit weird to me personally. I tried out both:

### Old

image


### New

image

File: crates/bevy_gizmos/src/primitives/dim3.rs (hunk 8 above)
Comment (RobWalt) [reply crates/bevy_gizmos/src/primitives/dim3.rs:493]: This was superseded by some other cleanup. Thanks for spotting it!

File: crates/bevy_gizmos/src/primitives/helpers.rs (hunk 11 above)
Comment (RobWalt) [reply crates/bevy_gizmos/src/primitives/helpers.rs:16]: Yeah, to be fair this seems to make only a minimal difference in terms of repetition. The idea here was to prevent the chance of making mistakes through typos, which would be higher if all the closures are inlined.

**Comment by RobWalt** [comment]
@Jondolf Examples are updated now and you can take a look at most primitives (except for the polygon ones)

File: crates/bevy_gizmos/src/primitives/dim3.rs [hunk 12]
```diff
@@ -0,0 +1,884 @@
...
+    pub fn len_segments(mut self, length: f32) -> Self {
+        self.len_segments = length;
+        self
+    }
+
+    /// Set the number of hinting axis used to hint the plane.
+    pub fn num_axis(mut self, num_axis: usize) -> Self {
+        self.num_axis = num_axis;
+        self
+    }
```
Comment (Jondolf) [inline crates/bevy_gizmos/src/primitives/dim3.rs:163]: I'm not quite sure what "hinting axis" and "hint the plane" mean. Also if this is plural, it should be "axes". I'd maybe also use `count` instead of `num`, like `axis_count` (and same for `num_segments`)

File: crates/bevy_gizmos/src/primitives/dim3.rs [hunk 13]
```diff
@@ -0,0 +1,884 @@
...
+    pub fn segments(mut self, segments: usize) -> Self {
+        self.num_segments = segments;
+        self
+    }
+
+    /// Set the length of segments used to hint the plane.
+    pub fn len_segments(mut self, length: f32) -> Self {
+        self.len_segments = length;
+        self
+    }
```
Comment (Jondolf) [inline crates/bevy_gizmos/src/primitives/dim3.rs:157]: This could maybe be fine, but `len` in Rust might give the impression of "the length of a collection" (e.g. `Vec::len`, in this case "the number of segments in this builder") instead of "math vector length".

Glam uses `length` for vector length, so `segment_length` might be clearer
```suggestion
    /// Set the length of the segments used to hint the plane.
    pub fn segment_length(mut self, length: f32) -> Self {
        self.segment_length = length;
        self
    }
```
(this would also match the word order of `segment_count` if we rename `num_segments` to that)

File: crates/bevy_gizmos/src/circles.rs [hunk 14]
```diff
@@ -4,20 +4,102 @@
...
+    ///         .ellipse_2d(Vec2::ZERO, Mat2::IDENTITY, 5., 1., Color::RED)
+    ///         .segments(64);
+    /// }
+    /// # bevy_ecs::system::assert_is_system(system);
+    /// ```
+    #[inline]
+    pub fn ellipse_2d(
+        &mut self,
+        position: Vec2,
+        rotation: Mat2,
```
Comment (Jondolf) [inline crates/bevy_gizmos/src/circles.rs:87]: No strong preference here, but `Mat2` might not be as user-friendly as just passing an `f32` angle in radians (like the bounding impls currently). If you don't know `Mat2::from_angle` exists, it'll probably be confusing to use.

My plan is to eventually add a `Rotation2d` type which would store a unit complex number (sine and cosine of rotation angle) and be efficient like `Mat2`. It'd implement `From<f32>` to create it from radians, so e.g. methods like this could take an `impl Into<Rotation2d>` and accept just radians (or a `Mat2` if desired) which are automatically converted to the rotation type

**Review by Jondolf — CHANGES_REQUESTED** [review]
A few more suggestions, mainly trivial fixes to the examples and some names

File: examples/3d/3d_gizmos.rs [hunk 15]
```diff
@@ -119,6 +182,168 @@ fn system(mut gizmos: Gizmos, mut my_gizmos: Gizmos<MyRoundGizmos>, time: Res<Ti
...
+        *segments_f = (*segments_f + 0.05).max(2.0);
+        segments.0 = segments_f.floor() as usize;
+    }
+    if keyboard.pressed(KeyCode::KeyH) {
+        *segments_f = (*segments_f - 0.05).max(2.0);
+        segments.0 = segments_f.floor() as usize;
+    }
+}
+
 fn rotate_camera(mut query: Query<&mut Transform, With<Camera>>, time: Res<Time>) {
```
[the remark this answers is not in the document]
Comment (RobWalt) [reply examples/3d/3d_gizmos.rs:347]: Not sure what you mean by that. Is it just for ordering reasons? Like grouping all `update` methods and all `draw` methods?

File: examples/3d/3d_gizmos.rs [hunk 16]
```diff
@@ -59,7 +120,9 @@ fn setup(
             Hold 'Left' or 'Right' to change the line width of straight gizmos\n\
             Hold 'Up' or 'Down' to change the line width of round gizmos\n\
             Press '1' or '2' to toggle the visibility of straight gizmos or round gizmos\n\
-            Press 'A' to show all AABB boxes",
+            Press 'A' to show all AABB boxes\n\
+            Press 'Up/Down' to cycle through primitives rendered with gizmos\n\
+            Press 'N/M' to decrease/increase the amount of segments in the primitives",
```
[the remark this answers is not in the document]
Comment (RobWalt) [reply examples/3d/3d_gizmos.rs]: Thanks for catching that!

File: examples/3d/3d_gizmos.rs (hunk 15 above)
[the remark this answers is not in the document]
Comment (RobWalt) [reply examples/3d/3d_gizmos.rs:351]: Ahh now I get it. Lol, sorry I'm blind 🙈

File: crates/bevy_gizmos/src/circles.rs [hunk 17]
```diff
@@ -4,20 +4,98 @@
...
     (0..segments + 1).map(move |i| {
         let angle = i as f32 * TAU / segments as f32;
-        Vec2::from(angle.sin_cos()) * radius
+        let (x, y) = angle.sin_cos();
+        Vec2::new(x, y) * half_size
     })
 }
 
 impl<'w, 's, T: GizmoConfigGroup> Gizmos<'w, 's, T> {
+    /// Draw a ellipse in 3D at `position` with the flat side facing `normal`.
```
Comment (Jondolf) [inline crates/bevy_gizmos/src/circles.rs:23]:
```suggestion
    /// Draw an ellipse in 3D at `position` with the flat side facing `normal`.
```

File: crates/bevy_gizmos/src/circles.rs [hunk 18]
```diff
@@ -4,20 +4,98 @@
...
+            gizmos: self,
+            position,
+            rotation,
+            half_size,
+            color,
+            segments: DEFAULT_CIRCLE_SEGMENTS,
+        }
+    }
+
+    /// Draw a ellipse in 2D.
```
Comment (Jondolf) [inline crates/bevy_gizmos/src/circles.rs:61]:
```suggestion
    /// Draw an ellipse in 2D.
```

File: examples/2d/2d_gizmos.rs [hunk 19]
```diff
@@ -1,29 +1,79 @@
...
+
 fn setup(mut commands: Commands, asset_server: Res<AssetServer>) {
     commands.spawn(Camera2dBundle::default());
     // text
     commands.spawn(TextBundle::from_section(
         "Hold 'Left' or 'Right' to change the line width of straight gizmos\n\
         Hold 'Up' or 'Down' to change the line width of round gizmos\n\
-        Press '1' or '2' to toggle the visibility of straight gizmos or round gizmos",
+        Press '1' or '2' to toggle the visibility of straight gizmos or round gizmos\n\
+        Press 'k/j' to cycle through primitives rendered with gizmos",
```
Comment (Jondolf) [inline examples/2d/2d_gizmos.rs:76]: Consistency
```suggestion
        Press '1' or '2' to toggle the visibility of straight gizmos or round gizmos\n\
        Press 'K' or 'J' to cycle through primitives rendered with gizmos",
```

File: examples/3d/3d_gizmos.rs [hunk 20]
```diff
@@ -59,7 +121,9 @@ fn setup(
             Hold 'Left' or 'Right' to change the line width of straight gizmos\n\
             Hold 'Up' or 'Down' to change the line width of round gizmos\n\
             Press '1' or '2' to toggle the visibility of straight gizmos or round gizmos\n\
-            Press 'A' to show all AABB boxes",
+            Press 'A' to show all AABB boxes\n\
+            Press 'k/j' to cycle through primitives rendered with gizmos\n\
+            Press 'h/l' to decrease/increase the amount of segments in the primitives",
```
Comment (Jondolf) [inline examples/3d/3d_gizmos.rs:126]: Consistency
```suggestion
            Press 'A' to show all AABB boxes\n\
            Press 'K' or 'J' to cycle through primitives rendered with gizmos\n\
            Press 'H' or 'L' to decrease/increase the amount of segments in the primitives",
```

**Review by Jondolf — APPROVED** [review]
A few more small doc issues, but I think this looks quite good overall! Huge thanks for all the patience :)

File: crates/bevy_gizmos/src/primitives/helpers.rs [hunk 21]
```diff
@@ -0,0 +1,94 @@
+use std::f32::consts::TAU;
+
+use bevy_math::{Mat2, Quat, Vec2, Vec3};
+use bevy_render::color::Color;
+
+use crate::prelude::{GizmoConfigGroup, Gizmos};
+
+// helpers - affine transform
+
+pub fn rotate_then_translate_2d(angle: f32, position: Vec2) -> impl Fn(Vec2) -> Vec2 {
```
Comment (alice-i-cecile) [inline crates/bevy_gizmos/src/primitives/helpers.rs:10]: I would really like docs for these public methods.

**Review by alice-i-cecile — APPROVED** [review]
A ton of code, but very straightforward. A nice demonstration of the utility of our new primitives.

I would *like* aggressive documentation, but it's not so bad that it can't be fixed in follow-up.

**Comment by alice-i-cecile** [comment]
Leaving this until Monday's merge train: if you have time to write docs this weekend, great! But I'm comfortable merging this either way.

File: crates/bevy_gizmos/src/primitives/helpers.rs (hunk 21 above)
Comment (RobWalt) [reply crates/bevy_gizmos/src/primitives/helpers.rs]: Good catch 🎯 They are exposed by the `pub(crate)` helper module but I guess I should've made them `pub(crate)` themselves as well. That's easier to read for the bevy devs which just jump there and are confused why they can't access it elsewhere.

I did some documentation there as well anyways
